swap/src/lib.rs

```diff
@@ -1,462 +1,307 @@
+            tx_in,
+            confidential,
+            unblinded,
         })
     }
 }
 
-pub struct Bob1 {
-    transaction: Transaction,
-    input_index_bob: usize,
-    amount_in_bob: ValueCommitment,
+#[derive(Debug)]
+struct UnblindedInput {
+    pub tx_in: TxIn,
+    pub confidential: ConfidentialTxOut,
+    pub unblinded: UnblindedTxOut,
 }
 
-impl Bob1 {
-    pub async fn compose(
-        &self,
-        signer: impl Future<Output = Result<Transaction>>,
-    ) -> Result<Message1> {
-        let transaction = signer.await?;
+#[derive(Debug)]
+pub struct Actor {
+    /// The inputs to cover for the trade.
+    inputs: Vec<UnblindedInput>,
+    /// The actors's address.
+    ///
+    /// Used for change as well as receive.
+    address: Address,
+    /// The ID of the asset the actor is receiving.
+    receive_asset: AssetId,
+    /// How much of the asset the actor is receiving.
+    receive_amount: Amount,
+}
 
-        Ok(Message1 { transaction })
+impl Actor {
+    pub fn new<C>(
+        secp: &Secp256k1<C>,
+        inputs: Vec<Input>,
+        address: Address,
+        receive_asset: AssetId,
+        receive_amount: Amount,
+    ) -> Result<Self>
+    where
+        C: Verification,
+    {
+        let inputs = inputs
+            .into_iter()
+            .map(|input| input.into_unblinded_input(secp))
+            .collect::<Result<Vec<_>, _>>()?;
+
+        Ok(Self {
+            inputs,
+            address,
+            receive_asset,
+            receive_amount,
+        })
     }
 
-    pub fn sign_with_key(&self, input_sk: &SecretKey) -> Result<Transaction> {
-        let secp = elements_fun::bitcoin::secp256k1::Secp256k1::new();
-
-        let input_pk_bob = SecpPublicKey::from_secret_key(&secp, &input_sk);
-
-        let mut transaction = self.transaction.clone();
-        transaction.input[self.input_index_bob]
-            .witness
-            .script_witness = {
-            let hash = hash160::Hash::hash(&input_pk_bob.serialize());
-            let script = Builder::new()
-                .push_opcode(opcodes::all::OP_DUP)
-                .push_opcode(opcodes::all::OP_HASH160)
-                .push_slice(&hash.into_inner())
-                .push_opcode(opcodes::all::OP_EQUALVERIFY)
-                .push_opcode(opcodes::all::OP_CHECKSIG)
-                .into_script();
-
-            let sighash = SigHashCache::new(&transaction).segwitv0_sighash(
-                self.input_index_bob,
-                &script,
-                self.amount_in_bob,
-                SigHashType::All,
-            );
-
-            let sig = secp.sign(&Message::from(sighash), &input_sk);
-
-            let mut serialized_signature = sig.serialize_der().to_vec();
-            serialized_signature.push(SigHashType::All as u8);
-
-            vec![serialized_signature, input_pk_bob.serialize().to_vec()]
-        };
+    pub fn calculate_change_amount(
+        &self,
+        other_receive_asset: AssetId,
```

Comment:
:question: Maybe `send_asset`?

Reply:
I am indifferent and lazy so I will just leave it at that :D